org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend

```diff
@@ -4850,73 +4850,102 @@ class SemanticHighlightingInformation {
 }
 
 /**
- * Returns a collection of calls from one symbol to another.
+ * The parameter of a `textDocument/prepareCallHierarchy` request.
  */
 @Beta
 @JsonRpcData
-class CallHierarchyParams extends TextDocumentPositionParams {
+class CallHierarchyPrepareParams  extends TextDocumentPositionParams {
 
-	/**
-	 * The direction of calls to provide.
-	 */
-	@NonNull
-	CallHierarchyDirection direction
+}
 
+/**
+ * The parameter of a `callHierarchy/incomingCalls` request.
+ */
+@Beta
+@JsonRpcData
+class CallHierarchyIncomingCallsParams {
+	CallHierarchyItem item
 }
 
 /**
- * Each {@code CallHierarchyCall} object defines a call from one {@code CallHierarchySymbol} to another.
+ * The parameter of a `callHierarchy/outgoingCalls` request.
  */
 @Beta
 @JsonRpcData
-class CallHierarchyCall {
+class CallHierarchyOutgoingCallsParams {
+	CallHierarchyItem item
+}
+
+/**
+ * Represents an incoming call, e.g. a caller of a method or constructor.
+ */
+@Beta
+@JsonRpcData
+class CallHierarchyIncomingCall {
 
 	/**
-	 * The source range of the reference. The range is a sub range of the {@link CallHierarchyCall#getFrom from} symbol range.
+	 * The item that makes the call.
 	 */
-	@NonNull
-	Range range
+	CallHierarchyItem from
 
 	/**
-	 * The symbol that contains the reference.
+	 * The range at which at which the calls appears. This is relative to the caller
+	 * denoted by [`this.from`](#CallHierarchyIncomingCall.from).
 	 */
-	@NonNull
-	CallHierarchySymbol from
+	Range[] fromRanges
+}
+
+/**
+ * Represents an outgoing call, e.g. calling a getter from a method or a method from a constructor etc.
+ */
+@Beta
+@JsonRpcData
+class CallHierarchyOutgoingCall {
 
 	/**
-	 * The symbol that is referenced.
+	 * The item that is called.
 	 */
-	@NonNull
-	CallHierarchySymbol to
+	CallHierarchyItem to
 
+	/**
+	 * The range at which this item is called. This is the range relative to the caller, e.g the item
+	 * passed to [`provideCallHierarchyOutgoingCalls`](#CallHierarchyItemProvider.provideCallHierarchyOutgoingCalls)
+	 * and not [`this.to`](#CallHierarchyOutgoingCall.to).
+	 */
+	Range[] fromRanges
 }
 
 /**
  * The result of a {@code textDocument/callHierarchy} request.
  */
 @Beta
 @JsonRpcData
-class CallHierarchySymbol {
+class CallHierarchyItem {
 
 	/**
-	 * The name of the symbol targeted by the call hierarchy request.
+	 * The name of the item targeted by the call hierarchy request.
 	 */
 	@NonNull
 	String name
 
 	/**
-	 * More detail for this symbol, e.g the signature of a function.
+	 * More detail for this item, e.g the signature of a function.
 	 */
 	String detail
 
 	/**
-	 * The kind of this symbol.
+	 * The kind of this item.
 	 */
 	@NonNull
 	SymbolKind kind
 
 	/**
-	 * URI of the document containing the symbol.
+	 * Tags for this item.
+	 */
+	SymbolTag[] tags
```

Comment:
`SymbolTag` is not defined here.

Reply:
SymbolTag is enum type. I saw other enums are all manually written. Is there any enum sample for Xtend definition?

Reply:
Looks like enum with constructor is not supported well in Xtend. https://bugs.eclipse.org/bugs/show_bug.cgi?id=403783

Reply:
You're right, the enums for LSP seem to be manual. I must have just remembered seeing them in [DebugProtocol.xtend](https://github.com/eclipse/lsp4j/blob/master/org.eclipse.lsp4j.debug/src/main/java/org/eclipse/lsp4j/debug/DebugProtocol.xtend).

Reply:
Yeah, i have a try on adding SymbolTag enum to Xtend, but it doesn't work.

Here are the problems:
- SymbolTag in LSP starts with 1, so the default enum definition doesn't work.
```java
enum SymbolTag {
  Deprecated
}
```
- @SerializedName can only specify an alias string for the enum type. But the value passed from LSP is a int number, the jsonrpc cannot convert it to the serializedName SymbolTag.
```java
enum SymbolTag {
  @SerializedName("1")
  Deprecated
}
```